File: post.py
from flask import Flask, request, jsonify

app = Flask(__name__)

# In-memory storage
posts = []
next_id = 1

# CREATE post
@app.route('/posts', methods=['POST'])
def create_post():
    global next_id
    data = request.get_json()

    if not data or 'title' not in data or 'content' not in data:
        return jsonify({'error': 'Invalid input'}), 400

    post = {
        'id': next_id,
        'title': data['title'],
        'content': data['content']
    }

    posts.append(post)
    next_id += 1

    return jsonify(post), 201


# READ all posts
@app.route('/posts', methods=['GET'])
def get_posts():
    return jsonify(posts), 200


# READ one post
@app.route('/posts/<int:post_id>', methods=['GET'])
def get_post(post_id):
    post = next((p for p in posts if p['id'] == post_id), None)

    if not post:
        return jsonify({'error': 'Post not found'}), 404

    return jsonify(post), 200


# UPDATE post
@app.route('/posts/<int:post_id>', methods=['PUT'])
def update_post(post_id):
    data = request.get_json()
    post = next((p for p in posts if p['id'] == post_id), None)

    if not post:
        return jsonify({'error': 'Post not found'}), 404

    post['title'] = data.get('title', post['title'])
    post['content'] = data.get('content', post['content'])

    return jsonify(post), 200


# DELETE post
@app.route('/posts/<int:post_id>', methods=['DELETE'])
def delete_post(post_id):
    global posts
    post = next((p for p in posts if p['id'] == post_id), None)

    if not post:
        return jsonify({'error': 'Post not found'}), 404

    posts = [p for p in posts if p['id'] != post_id]

    return jsonify({'message': 'Post deleted'}), 200


if __name__ == '__main__':
    app.run(debug=True)

File: post.c
#include "post.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#define MAX_POSTS 100

static Post posts[MAX_POSTS];
static int count = 0;
static int next_id = 1;

void init_posts() {
    count = 0;
    next_id = 1;
}

Post* create_post(const char* title, const char* content) {
    if (count >= MAX_POSTS) return NULL;

    Post* p = &posts[count++];
    p->id = next_id++;
    strncpy(p->title, title, sizeof(p->title));
    strncpy(p->content, content, sizeof(p->content));
    return p;
}

Post* get_post(int id) {
    for (int i = 0; i < count; i++)
        if (posts[i].id == id) return &posts[i];
    return NULL;
}

int update_post(int id, const char* title, const char* content) {
    Post* p = get_post(id);
    if (!p) return 0;

    strncpy(p->title, title, sizeof(p->title));
    strncpy(p->content, content, sizeof(p->content));
    return 1;
}

int delete_post(int id) {
    for (int i = 0; i < count; i++) {
        if (posts[i].id == id) {
            for (int j = i; j < count - 1; j++)
                posts[j] = posts[j + 1];
            count--;
            return 1;
        }
    }
    return 0;
}

char* get_all_posts_json() {
    char* buffer = malloc(4096);
    strcpy(buffer, "[");

    for (int i = 0; i < count; i++) {
        char temp[700];
        sprintf(temp,
            "{\"id\":%d,\"title\":\"%s\",\"content\":\"%s\"}%s",
            posts[i].id,
            posts[i].title,
            posts[i].content,
            (i < count - 1) ? "," : ""
        );
        strcat(buffer, temp);
    }

    strcat(buffer, "]");
    return buffer;
}
